win32ss/user/ntuser/window.c

```diff
@@ -3351,99 +3351,97 @@ NtUserFindWindowEx(HWND hwndParent,
    return Ret;
 }
 
-
-/*
- * @implemented
- */
-PWND FASTCALL UserGetAncestor(PWND Wnd, UINT Type)
+/* @implemented */
+PWND FASTCALL
+UserGetAncestor(_In_ PWND pWnd, _In_ UINT uType)
 {
-   PWND WndAncestor, Parent;
+    PWND WndAncestor, Parent, pwndMessage;
+    PDESKTOP pDesktop;
+    PWND pwndDesktop;
 
-   if (UserHMGetHandle(Wnd) == IntGetDesktopWindow())
-   {
-      return NULL;
-   }
+    pDesktop = pWnd->head.rpdesk;
+    ASSERT(pDesktop);
+    ASSERT(pDesktop->pDeskInfo);
 
-   switch (Type)
-   {
-      case GA_PARENT:
-         {
-            WndAncestor = Wnd->spwndParent;
-            break;
-         }
+    pwndDesktop = pDesktop->pDeskInfo->spwnd;
+    if (pWnd == pwndDesktop)
+        return NULL;
 
-      case GA_ROOT:
-         {
-            WndAncestor = Wnd;
-            Parent = NULL;
+    pwndMessage = pDesktop->spwndMessage;
+    if (pWnd == pwndMessage)
+        return NULL;
 
-            for(;;)
-            {
-               if(!(Parent = WndAncestor->spwndParent))
-               {
-                  break;
-               }
-               if(IntIsDesktopWindow(Parent))
-               {
-                  break;
-               }
+    Parent = pWnd->spwndParent;
+    if (!Parent)
+        return NULL;
 
-               WndAncestor = Parent;
-            }
-            break;
-         }
+    switch (uType)
+    {
+        case GA_PARENT:
+            return Parent;
 
-      case GA_ROOTOWNER:
-         {
-            WndAncestor = Wnd;
+        case GA_ROOT:
+            WndAncestor = pWnd;
+            if (Parent == pwndDesktop)
+                break;
 
-            for (;;)
+            do
             {
-               Parent = IntGetParent(WndAncestor);
+                if (Parent == pwndMessage)
+                    break;
 
-               if (!Parent)
-               {
-                  break;
-               }
+                WndAncestor = Parent;
+
+                pDesktop = Parent->head.rpdesk;
+                ASSERT(pDesktop);
+                ASSERT(pDesktop->pDeskInfo);
+
+                Parent = Parent->spwndParent;
+            } while (Parent != pDesktop->pDeskInfo->spwnd);
+            break;
 
-               WndAncestor = Parent;
+        case GA_ROOTOWNER:
+            for (PWND pwndNode = Parent; pwndNode; pwndNode = IntGetParent(pwndNode))
```

Comment:
I would suggest you initialize WndAncestor to something valid before this for-loop.

Reply:
Fixed.